### PR TITLE
Python 3 support, Windows installation fix, and others

### PR DESCRIPTION
**Pull request by uranusjr**
Gigantic bundle of updates.

## Unicode

All queries, statements and clauses now output Unicode objects (`str` on Python 3). This is done by adding

~~~python
from __future__ import unicode_literals
~~~

in `util.py`, and always using `''.format` and `%` inside `format` methods. **This can be backward-incompatible** if you have non-ASCII query strings, and does not use Unicode objects for your database connection (e.g. setting `use_unicode=False` in MySQL-Python), but should be good enough in most cases on non-Windows platform, and is better in the long run.

## Python 3 Support

Python 3 is now supported in all non-deprecated modules, and are test-covered. This is mainly handled by the `compat.py` module. All `print` statements are also converted to functions with

~~~python
from __future__ import print_function
~~~

This makes tests in deprecation modules fail, but I just skipped them (via configurations inside `runtests.py`). They can be converted to work, at least on Python 2, if you really want to, but I deemed it not worth the effort. I added a warning message in those modules (and `deprecation.rst`) to state that they are not tested anymore.

Many of the tests pass on both Python 2 and 3 after converting the print statements, but tests using `dict` may fail because hashing behaviours are different on Python 2 and 3, resulting in ordering change. I moved those tests to use unit-testing (using Nose), and use `OrderedDict` to promise ordering. They are inside `tests/`. The failing doctests are not removed, but converted to literal code blocks (using `::` so they are not picked up by Sphinx’s doctest extension), or skipped explicitly with `# doctest: +SKIP` ([documentation for this syntax](http://sphinx-doc.org/ext/doctest.html#directive-doctest)).

The original contents inside `tests/` are moved to `oldtests/`. I plan to add them back when I can find a good way to rewrite them. Will issue another pull request when that’s done.

## Testing

Tests are now handled by Nose, doctests included (with [sphinx-nose]). I maintain the sphinx-nose project. Tell me if there are any problems.

Parts of the documentation are slightly modified to make doctests compatible with nose. See commit message of b458cd0 for detailed information.

You will need to install some dependencies to run them:

    pip install -r requirements/test.txt

A simple `Makefile` is added to make things easier. To run the tests for the current Python version:

    make test

[Tox] is also used to provide cross-version testing support. Currently Python 2.7 and 3.4 are tested. I intended to add 2.6 to the support matrix, but found that there are some incompatibilities in `util.py`. They can be fixed quite easily, but I’m not sure whether this is needed. Drop a word if you do want 2.6 support.

To run Tox tests:

    pip install -r requirements/dev.txt
    make test-all

You will need both `python2.7` and `python3.4` somewhere in your `PATH`.

`.travis.yml` is also modified to use the new test structure (without tox).

## Misc. Bug Fixes

Installation on Windows was failing due to non-ASCII content inside `README.rst`. Fixed inside `setup.py`. (34112ad)

`changes.rst` was generating warnings about duplicate sqlmap references. Fixed with external hyperlink target. (ce449aa)

Some typo fixes in docs.

The `_is_iterable_not_str` function did not work as expected, and this became a problem on Python 3. This is fixed with other Python 3 fixes (921b031). The new implementation is:

~~~python
def _is_iterable_not_str(x):
    return (not isinstance(x, compat.class_types)
            and not isinstance(x, compat.string_types + (bytes,))
            and hasattr(x, '__iter__'))
~~~

The autoparam feature used a somewhat strange (IMO) way to check whether the parameter is a `param` class object. I changed the implementation to use proper class identity checking:

~~~python
# Inside _build_condition
if isinstance(v, compat.class_types) and issubclass(v, param):
    v = v(k)
~~~

[sphinx-nose]: https://github.com/uranusjr/saim-al-dahr
[Tox]: http://tox.readthedocs.org/en/latest/

### CODE REVIEW

**Comment by uranusjr** [comment]
I just added an edge case with Unicode and binary inputs. But I will need to test this on Windows tomorrow first, so please don’t merge this just yet.

**Comment by uranusjr** [comment]
All tests pass on Windows now.

**Comment by uranusjr** [comment]
Just added universal (Python 2 and 3) [wheel] support. To release a new version to PyPI, use

    make release

which is equivalent to

    python setup.py sdist bdist_wheel upload

One dependency is added to `requirements/dev.txt` for wheel-building support.

[wheel]: http://pythonwheels.com

**Comment by uranusjr** [comment]
Rebased and squashed the commits a bit.

**Comment by moskytw** [comment]
It's a great PR!

No, I don't think we need to support Python 2.6. I am checking the code.

File: MANIFEST.in [hunk 1]
```diff
@@ -1,6 +1,8 @@
 global-include README* LICENSE*
 graft examples
+graph oldtests
```
Comment (moskytw) [inline MANIFEST.in]: graph?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	$(MAKE) -C docs pickle >/dev/null 2>&1
+
+release: all
+	upload
+
+test: prepare-test
+	python runtests.py
+
+test-all: prepare-test
+	tox
```
Comment (moskytw) [inline Makefile]: nice!

File: docs/conf.py [hunk 3]
```diff
@@ -13,6 +13,8 @@
 
 import sys, os
 
+sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
```
Comment (moskytw) [inline docs/conf.py]: double dir name?

File: docs/make.bat [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+if "%1" == "pseudoxml" (
+	%SPHINXBUILD% -b pseudoxml %ALLSPHINXOPTS% %BUILDDIR%/pseudoxml
+	if errorlevel 1 exit /b 1
+	echo.
+	echo.Build finished. The pseudo-XML files are in %BUILDDIR%/pseudoxml.
+	goto end
+)
+
+:end
```
Comment (moskytw) [inline docs/make.bat]: I totally don't understand this. What a Windows stuff. But okay.

File: mosql/compat.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    integer_types = (int, long)
+    class_types = (type, types.ClassType)
+    text_type = unicode
+    binary_type = str
+
+
+try:
+    from itertools import izip
+except ImportError:
+    izip = zip
```
Comment (moskytw) [inline mosql/compat.py]: What a small module! nice.

File: mosql/util.py [hunk 6]
```diff
@@ -237,29 +241,36 @@ def __repr__(self):
...
 
 ___ = param
 
 # qualifier functions
 
 def _is_iterable_not_str(x):
-    return not isinstance(x, basestring) and hasattr(x, '__iter__')
+    return (not isinstance(x, compat.class_types)
+            and not isinstance(x, compat.string_types + (bytes,))
+            and hasattr(x, '__iter__'))
```
Comment (moskytw) [inline mosql/util.py]: I really wanna make it simpler ... can't we?

File: mosql/util.py [hunk 7]
```diff
@@ -237,29 +241,36 @@ def __repr__(self):
...
+
+def _coerce_str(x):
+    if isinstance(x, compat.binary_type):
+        x = x.decode('utf-8')
+    return x
```
Comment (moskytw) [inline mosql/util.py]: I am thinking just let binary fail. what do you think?

File: setup.py [hunk 8]
```diff
@@ -5,35 +5,37 @@
 
 import mosql
 
-import sys
-extra = {}
-if sys.version_info >= (3,):
-    extra['use_2to3'] = True
```
Comment (moskytw) [inline setup.py]: lol did I support Python 3?

File: setup.py [hunk 9]
```diff
@@ -5,35 +5,37 @@
...
+    url='http://mosql.mosky.tw/',
+    license='MIT',
+    classifiers=[
         'Development Status :: 5 - Production/Stable',
         'Environment :: Web Environment',
         'Intended Audience :: Developers',
         'License :: OSI Approved :: MIT License',
         'Operating System :: OS Independent',
         'Programming Language :: Python',
+        'Programming Language :: Python :: 2.7',
```
Comment (moskytw) [inline setup.py]: and Python 3 now I think.

File: tests/test_basic.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+from mosql.query import insert
+
+
+def test_simple_insert():
+    mosky = OrderedDict([
+        ('person_id', 'mosky'),
+        ('name', 'Mosky Liu'),
+    ])
+    exp = 'INSERT INTO "person" ("person_id", "name") VALUES (\'mosky\', \'Mosky Liu\')'
+    eq_(insert('person', mosky), exp)
```
Comment (moskytw) [inline tests/test_basic.py]: is eq same as just assert?

File: mosql/util.py (hunk 7 above)
Comment (moskytw) [inline mosql/util.py]: no. I think we should support this.

**Comment by moskytw** [comment]
looks great. just have some comments. :)

File: mosql/util.py (hunk 7 above)
Comment (uranusjr) [inline mosql/util.py]: How about a deprecation warning?

File: setup.py (hunk 8 above)
Comment (uranusjr) [inline setup.py]: You added this during the sprint last year IIRC. But the support was broken after 0.9.1.

File: tests/test_basic.py (hunk 10 above)
Comment (uranusjr) [inline tests/test_basic.py]: Yes, `eq_(a, b)` is essentially `assert a == b`. You can supply a third argument to customise the exception message, but it’s not used here.

File: docs/make.bat (hunk 4 above)
Comment (uranusjr) [inline docs/make.bat]: This is actually copied verbatim from Sphinx’s project template. But useful.

File: docs/conf.py (hunk 3 above)
Comment (uranusjr) [inline docs/conf.py]: I want to include the root project directory, so I can use `import mosql`. Previously you need to always `make -C docs html` to create docs, but I want to make this work with `cd docs && make html` because there’s no `make -C` on Windows.

~~~python
__file__ -> <project_root>/docs/conf.py
os.path.dirname(__file__) -> <project_root>/docs/
os.path.dirname(os.path.dirname(__file__)) -> <project_root>/
~~~

File: MANIFEST.in (hunk 1 above)
Comment (uranusjr) [inline MANIFEST.in]: Typo!

File: mosql/util.py (hunk 6 above)
Comment (uranusjr) [inline mosql/util.py]: I think the `+ (bytes,)` hack is droppable if we can work out all code paths to this function, and eliminate all possibility of a `bytes` instance being passed here (using `_coerce_str` or something). The extra type check is needed unless autoparam support is dropped.

File: docs/conf.py (hunk 3 above)
Comment (moskytw) [inline docs/conf.py]: I usually use `..` with `normpath`:

```python
>>> normpath(join('a/b/c', '../..'))
'a'
```

and if we are in a git repos, I will use this trick `git rev-parse --show-toplevel 2> /dev/null`.

File: docs/make.bat (hunk 4 above)
Comment (moskytw) [inline docs/make.bat]: oh, then could you comment about the source? a link to GitHub will be fine.

File: mosql/util.py (hunk 7 above)
Comment (moskytw) [inline mosql/util.py]: I am thinking ...

to keep this in Python 2, because MoSQL was a UTF-8-based project. We need to keep the basic backward compatibility.

just let binary fail in Python 3. in Python 3, you have no reason to pass a binary into here.

File: setup.py (hunk 8 above)
Comment (moskytw) [inline setup.py]: is it? lol

File: tests/test_basic.py (hunk 10 above)
Comment (moskytw) [inline tests/test_basic.py]: okay I see. usually I prefer to use simple assert.

Comment (uranusjr) [inline tests/test_basic.py]: Should I change this then?

File: mosql/util.py (hunk 7 above)
Comment (uranusjr) [inline mosql/util.py]: Sounds reasonable, except I have not idea how to make this work. If things work on Python 2, it would work on Python 3. Not the other way around.

File: docs/conf.py (hunk 3 above)
Comment (uranusjr) [inline docs/conf.py]: Don’t like the Git repo assumption. I’ll go with the `normpath` implementation.

Comment (moskytw) [inline docs/conf.py]: nice!

File: mosql/util.py (hunk 7 above)
Comment (moskytw) [inline mosql/util.py]: something like:

```python
if PY2:
    # for backward compatiblity
    def a():
        ...
else:
    # the perfect implementation
    def a():
        ...
```
```

File: tests/test_basic.py (hunk 10 above)
Comment (moskytw) [inline tests/test_basic.py]: It's just not my style, but it's good enough.

File: mosql/util.py (hunk 6 above)
Comment (moskytw) [inline mosql/util.py]: okay ... can we have two versions of this function? for py2 and py3. will it be simpler?

Comment (uranusjr) [inline mosql/util.py]: It will probably look like

~~~python
if PY2:
    def _is_iterable_not_str(x):
        return hasattr(x, '__iter__')  # The isinstance part can be dropped.
else:
    def _is_iterable_not_str(x):
        (not isinstance(x, type) and not isinstance(x, str) and hasattr(x, '__iter__')
~~~

Comment (uranusjr) [inline mosql/util.py]: Just had another idea. We can just call `iter` here and see what happens:

~~~python
def _is_iterable_not_str(x):
    if isintance(x, compat.string_types):
        return False
    try:
        iter(x)
    except TypeError:   # Not iterable.
        return False
    return True
~~~

Comment (moskytw) [inline mosql/util.py]: no don't use try for this. that will be heavy. this function should be super lightweight.

Comment (uranusjr) [inline mosql/util.py]: Fair enough. There’s no good way around this then.

File: mosql/__init__.py [hunk 11]
```diff
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 
-__author__ = 'Mosky <http://mosky.tw>'
-__version__ = '0.10'
+VERSION = (0, 10,)
 
+__author__ = 'Mosky <http://mosky.tw>'
+__version__ = '.'.join(str(v) for v in VERSION)
```
Comment (moskytw) [inline mosql/__init__.py]: why?

File: mosql/util.py [hunk 12]
```diff
@@ -237,80 +241,108 @@ def __repr__(self):
...
+qualifier = _qualifier
+'''A decorator which makes all items in an `iterable` apply a qualifier
+function, `f`, or simply apply the qualifier function to the input if the input
+is not an `iterable`.
+
+The `iterable` here means the iterable except string.
+
+It also makes a qualifier function returns the input without changes if the
+input is an instance of :class:`raw`.
+'''
```
Comment (moskytw) [inline mosql/util.py]: nice

File: mosql/__init__.py (hunk 11 above)
Comment (uranusjr) [inline mosql/__init__.py]: Because

    >>> import mosql
    >>> print(mosql.__version__)
    0.10
    >>> mosql.__version__ > '0.9'
    False
    >>> print(mosql.VERSION)
    (0, 10)
    >>> mosql.VERSION > (0, 9)
    True

I stole the idea from [Django]’s implementation, but there are probably many projects using this pattern.

[Django]: https://github.com/django/django/blob/master/django/__init__.py
